src/apps/intel_mp/intel_mp.lua

```diff
@@ -0,0 +1,774 @@
+MDIC      0x00020 -                 RW MDI Control
+MDICNFG   0x00E04 -                 RW MDI Configuration
+RCTL      0x00100 -                 RW RX Control
+RPTHC     0x04104 -                 RC Rx Packets to host count
+SW_FW_SYNC 0x05b5c -                RW Software Firmware Synchronization
+TCTL      0x00400 -                 RW TX Control
+TCTL_EXT  0x00400 -                 RW Extended TX Control
+]],
+   txq = [[
+TDBAL  0xe000 +0x40*0..7            RW Tx Descriptor Base Low
+TDBAH  0xe004 +0x40*0..7            RW Tx Descriptor Base High
+TDLEN  0xe008 +0x40*0..7            RW Tx Descriptor Ring Length
+TDH    0xe010 +0x40*0..7            RO Tx Descriptor Head
+TDT    0xe018 +0x40*0..7            RW Tx Descriptor Tail
+TXDCTL 0xe028 +0x40*0..7            RW Tx Descriptor Control Queue
+TXCTL  0xe014 +0x40*0..7            RW Tx DCA CTRL Register Queue
+]]
+}
+reg.i210 = {
+   inherit = "1000BaseX",
+   singleton = [[
+EEMNGCTL  0x12030 -            RW Manageability EEPROM-Mode Control Register
+EEC       0x12010 -            RW EEPROM-Mode Control Register
+]]
+}
+reg.i350 = {
+   array = [[
+ALLRQDPC  0xC130 +0x40*4..7         RW Receive Queue drop packet count Register
+]],
+   inherit = "1000BaseX",
+   singleton = [[
+EEMNGCTL  0x01010 -            RW Manageability EEPROM-Mode Control Register
+EEC       0x00010 -            RW EEPROM-Mode Control Register
+]]
+}
+reg["0x1521"] = { inherit = "i350" }
+reg["0x1533"] = { inherit = "i210" }
+reg["0x157b"] = { inherit = "i210" }
+reg["0x10fb"] = { inherit = "82599ES" }
+
+local Intel = { }
+function Intel:new (arg)
+   local conf = config.parse_app_arg(arg)
+   local self = setmetatable({
+      r = {},
+      pciaddress = conf.pciaddr,
+      path = pci.path(conf.pciaddr),
+      ndesc = conf.ndescriptors or 2048,
+      txq = conf.txq,
+      rxq = conf.rxq,
+      rssseed = conf.rssseed or 314159,
+      linkup_attempts = conf.linkup_attempts or 60,
+      wait_for_link = conf.wait_for_link or false
+   }, {__index = self})
+
+   local vendor = lib.firstline(self.path .. "/vendor")
+   local device = lib.firstline(self.path .. "/device")
+   assert(vendor == '0x8086', "unsupported nic")
+   self.max_q = self.max_q_by_dev[device]
+
+   assert(self.max_q, "Unsupported Intel NIC")
+
+   -- Setup device access
+   self.base, self.fd = pci.map_pci_memory_unlocked(self.pciaddress, 0)
+   self.master = self.fd:flock("ex, nb")
+
+   self:load_registers(device)
+
+   self:init()
+   self.fd:flock("sh")
+   self:init_tx_q()
+   self:init_rx_q()
+   return self
+end
+
+function Intel:disable_interrupts ()
+   self.r.EIMC(0xffffffff)
+end
+function Intel:init_rx_q ()
+   if not self.rxq then return end
+   assert((self.rxq >=0) and (self.rxq < self.max_q),
+   "rxqueue must be in 0.." .. self.max_q-1)
+   assert((self.ndesc %128) ==0,
+   "ndesc must be a multiple of 128 (for Rx only)")	-- see 7.1.4.5
+
+   self.rxpackets = ffi.new("struct packet *[?]", self.ndesc)
```

Comment:
Seems cleaner to me to not invoke the C name directly, in case the ever definition changes. Also, if `desq_t` was defined module-wide, then the JIT would see one ctype instead of different ctypes for each app instance.